File: typescript/src/token-service.ts
/**
 * `issueToken`.
 *
 * Token issuance refuses non-issuable statuses (suspended, revoked, expired,
 * pending) and builds the LIC1 payload with the required + optional claim
 * set.
 *
 * Semantics:
 *
 *   1. Loads the *active* signing key for the license's scope and the
 *      caller-requested `alg`. If none exists and the license has a scope,
 *      falls back to the global (scope_id = null) key — this mirrors the
 *      PHP reference's "scoped → global" fallback so operators can run a
 *      single global signer until they're ready to split per-scope keys.
 *
 *   2. Refuses to issue for non-usable statuses. `effectiveStatus` is used
 *      so that a license still within its grace window issues a token with
 *      `status: 'grace'` without requiring the caller to have called `tick`
 *      first. `pending` licenses cannot issue — activation (the first
 *      `registerUsage`) must have already happened.
 *
 *   3. Claims are built to match the LIC1 payload spec exactly — `jti` is a
 *      fresh uuidv7 per token, timestamps are unix seconds (iat/nbf/exp),
 *      `scope` is the scope slug (empty string for global), `status` is the
 *      effective status, and optional claims (`force_online_after`,
 *      `entitlements`, `meta`) are OMITTED (not set to null) when unused so
 *      the canonical JSON is minimal.
 *
 *   4. `force_online_after` can come from (a) the caller-supplied override,
 *      (b) `license.meta.force_online_after_sec` (set by
 *      `createLicenseFromTemplate`), or (c) be absent entirely. It's
 *      expressed as an absolute unix-seconds deadline relative to `iat`, so
 *      clients can compare directly against their clock.
 *
 *   5. `entitlements` come from `license.meta.entitlements` (snapshotted at
 *      creation time by the template flow). Caller `entitlements` in opts
 *      wins if provided.
 *
 *   6. The signing key's private PEM is decrypted in-frame using the
 *      passphrase, the signature backend signs the canonical signing input,
 *      and the plaintext handle is NOT cached — every issue re-decrypts.
 *      This is a deliberate trade: a warm cache would save ~100ms/token
 *      but widens the memory-exposure window; the ~100ms is negligible
 *      versus network/DB round-trips in a real issuer service.
 *
 * Non-goals for 5.5:
 *   - Refresh/rotation of tokens (that's a caller-orchestration concern).
 *   - Batch issuance — callers loop.
 *   - A `refreshToken(oldToken, ...)` API — clients simply re-call issue.
 */

import type { SignatureBackend } from './crypto/types.ts';
import { errors } from './errors.ts';
import type { Clock } from './id.ts';
import { newUuidV7 } from './id.ts';
import { KeyHierarchy } from './key-hierarchy.ts';
import { encode as encodeLic1, type LIC1Header } from './lic1.ts';
import { effectiveStatus } from './lifecycle.ts';
import type { Storage } from './storage/types.ts';
import type {
  JSONValue,
  KeyAlg,
  License,
  LicenseKey,
  LicenseScope,
  LicenseUsage,
} from './types.ts';

export interface IssueTokenInput {
  /** The license being validated. Status is re-resolved via
   *  `effectiveStatus(now)` — caller does NOT need to have called `tick`
   *  first to promote active→grace. */
  readonly license: License;
  /** The active usage row (seat) this token is scoped to. Its fingerprint
   *  becomes the `usage_fingerprint` claim and binds the token to a device. */
  readonly usage: LicenseUsage;
  /** Token time-to-live in seconds. Required — no silent default so callers
   *  are forced to decide. Typical: 3600–86400 for online-gate'd apps,
   *  up to 30 days for offline-first. */
  readonly ttlSeconds: number;
  /** Algorithm selector for the signing key. The issuer picks the active
   *  key matching `(scope_id, role=signing, alg)`. */
  readonly alg: KeyAlg;
  /** Passphrase for the active signing key's encrypted PKCS8 blob. Provided
   *  per-call (from KMS/env/HSM) rather than cached in-memory. */
  readonly signingPassphrase: string;
  /** Optional override: absolute unix-seconds deadline by which the client
   *  MUST re-check with the server. When omitted, falls back to
   *  `license.meta.force_online_after_sec` relative to `iat` (if set).
   *  Pass `null` to explicitly omit even when the license meta has one. */
  readonly forceOnlineAfter?: number | null;
  /** Optional entitlements override — wins over `license.meta.entitlements`.
   *  Pass `null` to explicitly omit both. */
  readonly entitlements?: Readonly<Record<string, JSONValue>> | null;
  /** Optional freeform meta to include in the claim set. Not merged with
   *  the license's meta — this is a token-scoped claim only. */
  readonly meta?: Readonly<Record<string, JSONValue>>;
}

export interface IssueTokenResult {
  /** The encoded LIC1 token — safe to hand to a client for offline use. */
  readonly token: string;
  /** The signing key used (`kid`, `alg`) — handy for logging. */
  readonly kid: string;
  /** Unix seconds of `iat/nbf` (they're always equal for fresh issuance). */
  readonly iat: number;
  /** Unix seconds of `exp`. */
  readonly exp: number;
  /** Per-token uuidv7 — also the `jti` claim. */
  readonly jti: string;
}

/**
 * Issue a new LIC1 token for `(license, usage)` signed by the active
 * signing key for the license's scope. Does not persist anything — token
 * issuance is stateless at the storage layer (the audit row for "token
 * issued" is a future concern; we only audit *lifecycle* transitions, not
 * every token emission).
 */
export async function issueToken(
  storage: Storage,
  clock: Clock,
  backends: ReadonlyMap<KeyAlg, SignatureBackend>,
  input: IssueTokenInput,
): Promise<IssueTokenResult> {
  if (input.ttlSeconds <= 0) {
    throw errors.fingerprintRejected(`issueToken.ttlSeconds must be > 0 (got ${input.ttlSeconds})`);
  }
  if (input.usage.license_id !== input.license.id) {
    throw errors.tokenMalformed(
      `usage.license_id (${input.usage.license_id}) does not belong to license ${input.license.id}`,
    );
  }
  if (input.usage.status !== 'active') {
    throw errors.tokenMalformed(
      `usage ${input.usage.id} is not active (status=${input.usage.status})`,
    );
  }

  // Resolve status at issuance time. `effectiveStatus` promotes
  // active→grace when now is between expires_at and grace_until, and
  // active→expired once grace has passed.
  const nowIso = clock.nowIso();
  const status = effectiveStatus(input.license, nowIso);
  if (status === 'pending') {
    throw errors.tokenMalformed(
      `license ${input.license.id} is still pending — activate via registerUsage first`,
    );
  }
  if (status === 'suspended') throw errors.licenseSuspended();
  if (status === 'revoked') throw errors.licenseRevoked();
  if (status === 'expired') throw errors.licenseExpired();
  // Reachable: 'active' | 'grace'.

  // Load the active signing key. Scope-preferential: try scoped first, fall
  // back to global if the license has a scope but no scoped signer exists.
  const signing = await findActiveSigningKey(storage, input.license.scope_id, input.alg);
  if (signing === null) {
    throw errors.unknownKid(
      input.license.scope_id === null
        ? `no active global signing key for alg=${input.alg}`
        : `no active signing key for scope=${input.license.scope_id} alg=${input.alg} (and no global fallback)`,
    );
  }
  // Defense-in-depth scope drift guard: `findActiveSigningKey` already filters
  // `WHERE scope_id = <license.scope_id> OR (scope_id IS NULL AND no scoped
  // match)`, but we re-assert here so a DB-integrity bug that returns a key
  // from scope B for a license in scope A can't sign cross-tenant tokens.
  // Valid signers are either the license's own scope or the global fallback.
  if (signing.scope_id !== input.license.scope_id && signing.scope_id !== null) {
    throw errors.unknownKid(
      `signing key ${signing.kid} scope (${signing.scope_id}) does not match license scope (${input.license.scope_id ?? 'global'})`,
    );
  }

  // Resolve scope slug for the `scope` claim. Empty string means global.
  // Lookup only when the license has a scope.
  const scopeSlug =
    input.license.scope_id === null ? '' : await resolveScopeSlug(storage, input.license.scope_id);

  const iat = unixSeconds(nowIso);
  const exp = iat + input.ttlSeconds;
  const jti = newUuidV7(clock);

  // Build the payload in canonical shape. Optional claims are OMITTED
  // (not nulled) when unused to keep the canonical JSON minimal.
  const payload: Record<string, JSONValue> = {
    jti,
    iat,
    nbf: iat,
    exp,
    scope: scopeSlug,
    license_id: input.license.id,
    usage_id: input.usage.id,
    usage_fingerprint: input.usage.fingerprint,
    status,
    max_usages: input.license.max_usages,
  };

  const forceOnlineAfter = resolveForceOnlineAfter(input, iat);
  if (forceOnlineAfter !== null) payload.force_online_after = forceOnlineAfter;

  const entitlements = resolveEntitlements(input);
  if (entitlements !== null) payload.entitlements = entitlements as JSONValue;

  if (input.meta !== undefined) payload.meta = input.meta as JSONValue;

  // Sign. The handle is scoped to this frame and discarded; no cache.
  const hierarchy = new KeyHierarchy({
    store: {
      // Minimal adapter — importSigningPrivate only calls findByKid + one
      // of the backends. A thin inline store is fine.
      async put() {
        throw new Error('unreachable');
      },
      async get() {
        return null;
      },
      async findByKid(kid: string) {
        return kid === signing.kid ? signing : null;
      },
      async list() {
        return [signing];
      },
      async update() {
        throw new Error('unreachable');
      },
    },
    backends,
    clock,
  });
  const { handle } = await hierarchy.importSigningPrivate(signing.kid, input.signingPassphrase);

  const backend = backends.get(input.alg);
  if (!backend) throw errors.unsupportedAlgorithm(input.alg);

  const header: LIC1Header = {
    v: 1,
    typ: 'lic',
    alg: input.alg,
    kid: signing.kid,
  };
  const token = await encodeLic1({ header, payload, privateKey: handle, backend });

  return { token, kid: signing.kid, iat, exp, jti };
}

// ---------- internals ----------

async function findActiveSigningKey(
  storage: Storage,
  scope_id: License['scope_id'],
  alg: KeyAlg,
): Promise<LicenseKey | null> {
  // Scoped first. Walk pages because realistic counts are tiny but the
  // adapter interface forces us to respect paging.
  if (scope_id !== null) {
    const scoped = await findFirstActive(storage, { scope_id, alg });
    if (scoped !== null) return scoped;
  }
  // Global fallback (or direct query when scope_id was null to begin with).
  return findFirstActive(storage, { scope_id: null, alg });
}

async function findFirstActive(
  storage: Storage,
  filter: { scope_id: LicenseScope['id'] | null; alg: KeyAlg },
): Promise<LicenseKey | null> {
  let cursor: string | null | undefined;
  while (true) {
    const page = await storage.listKeys(
      { scope_id: filter.scope_id, alg: filter.alg, role: 'signing', state: 'active' },
      cursor === null || cursor === undefined ? { limit: 50 } : { limit: 50, cursor },
    );
    if (page.items.length > 0) return page.items[0] ?? null;
    if (page.cursor === null) return null;
    cursor = page.cursor;
  }
}

async function resolveScopeSlug(
  storage: Storage,
  scope_id: NonNullable<License['scope_id']>,
): Promise<string> {
  const scope = await storage.getScope(scope_id);
  if (scope === null) {
    throw errors.tokenMalformed(`license references unknown scope ${scope_id}`);
  }
  return scope.slug;
}

/** Cap at ~10 years past `iat`. Longer horizons are almost certainly
 *  misconfigured meta, and `iat + very_large_number` is a silent DoS against
 *  the "must re-check with server" guarantee. */
const FORCE_ONLINE_MAX_SEC = 10 * 365 * 24 * 3600;

/** Validate a candidate `force_online_after` deadline. Already absolute
 *  (unix seconds). Must be a finite non-negative safe integer within the
 *  capped horizon and not already past `iat`. */
function validateForceOnlineDeadline(candidate: number, iat: number, context: string): number {
  if (!Number.isFinite(candidate) || !Number.isSafeInteger(candidate)) {
    throw errors.fingerprintRejected(`${context} must be a finite integer (got ${candidate})`);
  }
  if (candidate < iat) {
    throw errors.fingerprintRejected(`${context} (${candidate}) is in the past (iat=${iat})`);
  }
  if (candidate > iat + FORCE_ONLINE_MAX_SEC) {
    throw errors.fingerprintRejected(
      `${context} (${candidate}) exceeds max horizon (${iat + FORCE_ONLINE_MAX_SEC})`,
    );
  }
  return candidate;
}

function resolveForceOnlineAfter(input: IssueTokenInput, iat: number): number | null {
  if (input.forceOnlineAfter !== undefined) {
    if (input.forceOnlineAfter === null) return null;
    return validateForceOnlineDeadline(input.forceOnlineAfter, iat, 'forceOnlineAfter');
  }
  const meta = input.license.meta as Record<string, unknown>;
  const fromMeta = meta.force_online_after_sec;
  // `force_online_after_sec` in meta is a RELATIVE duration (seconds from
  // now), not an absolute deadline — this is the shape `createLicenseFromTemplate`
  // writes. Validate the relative value, then convert to absolute.
  if (typeof fromMeta !== 'number') return null;
  if (!Number.isFinite(fromMeta) || !Number.isSafeInteger(fromMeta) || fromMeta <= 0) {
    throw errors.fingerprintRejected(
      `license.meta.force_online_after_sec must be a positive integer (got ${fromMeta})`,
    );
  }
  if (fromMeta > FORCE_ONLINE_MAX_SEC) {
    throw errors.fingerprintRejected(
      `license.meta.force_online_after_sec (${fromMeta}) exceeds max horizon (${FORCE_ONLINE_MAX_SEC})`,
    );
  }
  return iat + fromMeta;
}

function resolveEntitlements(input: IssueTokenInput): Readonly<Record<string, JSONValue>> | null {
  if (input.entitlements !== undefined) return input.entitlements;
  const meta = input.license.meta as Record<string, unknown>;
  const fromMeta = meta.entitlements;
  if (fromMeta !== undefined && fromMeta !== null && typeof fromMeta === 'object') {
    return fromMeta as Readonly<Record<string, JSONValue>>;
  }
  return null;
}

function unixSeconds(iso: string): number {
  return Math.floor(Date.parse(iso) / 1000);
}

File: licensing/token_service.go
package licensing

// Token issuance — Go port of
// typescript/packages/core/src/token-service.ts.

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"time"
)

// IssueTokenInput carries the caller-supplied parameters for issuing a
// LIC1 token.
type IssueTokenInput struct {
	ForceOnlineAfter OptIntOverride
	Entitlements     OptEntitlements
	License          *License
	Usage            *LicenseUsage
	Meta             map[string]any
	// TransparencyHook is an optional callback fired after a token is
	// successfully signed. It receives the token's identifying metadata
	// + a SHA-256 hash of the full wire-token bytes; operators MAY mirror
	// this to an externally-verifiable append-only store (S3 with object
	// lock, AWS QLDB, immudb, a managed CT-style log) so a stolen-key
	// attacker who mints tokens cannot do so without leaving a trail
	// on the operator's transparency vendor.
	//
	// The hook is fire-and-forget: any retry / async / error-surfacing
	// concern lives in the operator's wrapper. The token is already
	// signed and returned to the caller by the time the hook fires;
	// hook failures do NOT fail the issuance.
	//
	// Leave nil to disable; the hook costs ~zero when unset.
	TransparencyHook  TransparencyHook
	Alg               KeyAlg
	SigningPassphrase string
	TTLSeconds        int
}

// TransparencyHook is the signature of the post-issue callback. The
// event carries enough metadata to identify the token and verify its
// integrity against an external log, without leaking the token bytes
// themselves.
type TransparencyHook func(event TokenIssuedEvent)

// TokenIssuedEvent is the payload passed to TransparencyHook. All
// fields are non-empty for a successful issue. TokenSHA256 is the
// lowercase-hex SHA-256 of the full wire-token string (i.e. the same
// bytes the consumer receives), 64 chars.
type TokenIssuedEvent struct {
	Jti         string
	LicenseID   string
	UsageID     string
	Kid         string
	TokenSHA256 string
	Iat         int64
	Exp         int64
}

// OptIntOverride distinguishes "not supplied" from "explicitly set".
type OptIntOverride struct {
	Value *int64
	Set   bool
}

// OptEntitlements distinguishes "not supplied" from "explicitly set".
type OptEntitlements struct {
	Value map[string]any
	Set   bool
}

// IssueTokenResult is the return value of IssueToken.
type IssueTokenResult struct {
	Token string
	Kid   string
	Jti   string
	Iat   int64
	Exp   int64
}

// IssueToken builds and signs a LIC1 token for (license, usage). It does
// not persist anything — token issuance is stateless at the storage layer.
func IssueToken(
	storage Storage, clock Clock,
	backends *AlgorithmRegistry,
	input IssueTokenInput,
) (*IssueTokenResult, error) {
	if input.TTLSeconds <= 0 {
		return nil, newError(CodeFingerprintRejected,
			fmt.Sprintf("issueToken.ttlSeconds must be > 0 (got %d)", input.TTLSeconds), nil)
	}
	if input.Usage.LicenseID != input.License.ID {
		return nil, newError(CodeTokenMalformed,
			fmt.Sprintf("usage.license_id (%s) does not belong to license %s",
				input.Usage.LicenseID, input.License.ID), nil)
	}
	if input.Usage.Status != UsageStatusActive {
		return nil, newError(CodeTokenMalformed,
			fmt.Sprintf("usage %s is not active (status=%s)",
				input.Usage.ID, input.Usage.Status), nil)
	}

	nowISO := clock.NowISO()
	status := EffectiveStatus(input.License, nowISO)
	switch status {
	case LicenseStatusActive, LicenseStatusGrace:
		// OK — issuable.
	case LicenseStatusPending:
		return nil, newError(CodeTokenMalformed,
			fmt.Sprintf("license %s is still pending — activate via registerUsage first",
				input.License.ID), nil)
	case LicenseStatusSuspended:
		return nil, newError(CodeLicenseSuspended, "license is suspended", nil)
	case LicenseStatusRevoked:
		return nil, newError(CodeLicenseRevoked, "license is revoked", nil)
	case LicenseStatusExpired:
		return nil, newError(CodeLicenseExpired, "license has expired", nil)
	default:
		return nil, newError(CodeTokenMalformed,
			fmt.Sprintf("unknown effective status: %s", status), nil)
	}

	// Load the active signing key: scoped first, global fallback.
	signing, err := findActiveSigningKey(storage, input.License.ScopeID, input.Alg)
	if err != nil {
		return nil, err
	}
	if signing == nil {
		msg := fmt.Sprintf("no active global signing key for alg=%s", input.Alg)
		if input.License.ScopeID != nil {
			msg = fmt.Sprintf(
				"no active signing key for scope=%s alg=%s (and no global fallback)",
				*input.License.ScopeID, input.Alg)
		}
		return nil, newError(CodeUnknownKid, msg, nil)
	}

	// Defense-in-depth: key scope must match license scope or be global.
	if signing.ScopeID != nil {
		licenseScopeMatch := input.License.ScopeID != nil && *signing.ScopeID == *input.License.ScopeID
		if !licenseScopeMatch {
			scopeLabel := "global"
			if input.License.ScopeID != nil {
				scopeLabel = *input.License.ScopeID
			}
			return nil, newError(CodeUnknownKid,
				fmt.Sprintf("signing key %s scope (%s) does not match license scope (%s)",
					signing.Kid, *signing.ScopeID, scopeLabel), nil)
		}
	}

	// Resolve scope slug for the `scope` claim.
	scopeSlug := ""
	if input.License.ScopeID != nil {
		scope, err := storage.GetScope(*input.License.ScopeID)
		if err != nil {
			return nil, err
		}
		if scope == nil {
			return nil, newError(CodeTokenMalformed,
				fmt.Sprintf("license references unknown scope %s", *input.License.ScopeID), nil)
		}
		scopeSlug = scope.Slug
	}

	iat := unixSeconds(nowISO)
	exp := iat + int64(input.TTLSeconds)
	jti := NewUUIDv7()

	// Build payload.
	payload := LIC1Payload{
		"jti":               jti,
		"iat":               iat,
		"nbf":               iat,
		"exp":               exp,
		"scope":             scopeSlug,
		"license_id":        input.License.ID,
		"usage_id":          input.Usage.ID,
		"usage_fingerprint": input.Usage.Fingerprint,
		"status":            string(status),
		"max_usages":        int64(input.License.MaxUsages),
	}

	forceOnlineAfter, err := resolveForceOnlineAfter(input, iat)
	if err != nil {
		return nil, err
	}
	if forceOnlineAfter != nil {
		payload["force_online_after"] = *forceOnlineAfter
	}

	entitlements := resolveEntitlements(input)
	if entitlements != nil {
		payload["entitlements"] = entitlements
	}

	if input.Meta != nil {
		payload["meta"] = input.Meta
	}

	// Import signing key's private material. Use a minimal KeyStore adapter
	// that only serves the one key we already loaded.
	hierarchy, err := NewKeyHierarchy(KeyHierarchyOptions{
		Store:    &singleKeyStore{key: signing},
		Registry: backends,
		Clock:    clock,
	})
	if err != nil {
		return nil, err
	}
	_, handle, err := hierarchy.ImportSigningPrivate(signing.Kid, input.SigningPassphrase)
	if err != nil {
		return nil, err
	}

	backend, err := backends.Get(input.Alg)
	if err != nil {
		return nil, err
	}

	header := LIC1Header{V: 1, Typ: "lic", Alg: input.Alg, Kid: signing.Kid}
	token, err := Encode(EncodeOptions{
		Header:     header,
		Payload:    payload,
		PrivateKey: handle,
		Backend:    backend,
	})
	if err != nil {
		return nil, err
	}

	// Fire the transparency hook. Hash the full wire-token bytes —
	// that's what an external log would record and what a third party
	// would compare against the operator's local audit log to detect a
	// stolen-key issuance. SHA-256 of UTF-8 bytes; 64-char lowercase hex.
	if input.TransparencyHook != nil {
		sum := sha256.Sum256([]byte(token))
		input.TransparencyHook(TokenIssuedEvent{
			Jti:         jti,
			LicenseID:   input.License.ID,
			UsageID:     input.Usage.ID,
			Kid:         signing.Kid,
			Iat:         iat,
			Exp:         exp,
			TokenSHA256: hex.EncodeToString(sum[:]),
		})
	}

	return &IssueTokenResult{
		Token: token,
		Kid:   signing.Kid,
		Iat:   iat,
		Exp:   exp,
		Jti:   jti,
	}, nil
}

// ---------- internals ----------

// findActiveSigningKey finds the active signing key for a scope and alg.
// Tries scoped first, falls back to global.
func findActiveSigningKey(storage Storage, scopeID *string, alg KeyAlg) (*LicenseKey, error) {
	if scopeID != nil {
		key, err := findFirstActiveKey(storage, scopeID, alg)
		if err != nil {
			return nil, err
		}
		if key != nil {
			return key, nil
		}
	}
	// Global fallback (or direct query when scopeID was nil).
	return findFirstActiveKey(storage, nil, alg)
}

func findFirstActiveKey(storage Storage, scopeID *string, alg KeyAlg) (*LicenseKey, error) {
	role := RoleSigning
	state := StateActive
	cursor := ""
	for {
		page, err := storage.ListKeys(LicenseKeyFilter{
			ScopeID:    scopeID,
			ScopeIDSet: true,
			Alg:        &alg,
			Role:       &role,
			State:      &state,
		}, PageRequest{Limit: 50, Cursor: cursor})
		if err != nil {
			return nil, err
		}
		if len(page.Items) > 0 {
			return &page.Items[0], nil
		}
		if page.Cursor == "" {
			return nil, nil
		}
		cursor = page.Cursor
	}
}

const forceOnlineMaxSec int64 = 10 * 365 * 24 * 3600

func resolveForceOnlineAfter(input IssueTokenInput, iat int64) (*int64, error) {
	if input.ForceOnlineAfter.Set {
		if input.ForceOnlineAfter.Value == nil {
			return nil, nil
		}
		v := *input.ForceOnlineAfter.Value
		if err := validateForceOnlineDeadline(v, iat, "forceOnlineAfter"); err != nil {
			return nil, err
		}
		return &v, nil
	}
	// Fall back to license.meta.force_online_after_sec (relative duration).
	if input.License.Meta == nil {
		return nil, nil
	}
	raw, ok := input.License.Meta["force_online_after_sec"]
	if !ok {
		return nil, nil
	}
	sec, ok := toInt64(raw)
	if !ok {
		return nil, newError(CodeFingerprintRejected,
			fmt.Sprintf("license.meta.force_online_after_sec must be a positive integer (got %v)", raw), nil)
	}
	if sec <= 0 {
		return nil, newError(CodeFingerprintRejected,
			fmt.Sprintf("license.meta.force_online_after_sec must be a positive integer (got %d)", sec), nil)
	}
	if sec > forceOnlineMaxSec {
		return nil, newError(CodeFingerprintRejected,
			fmt.Sprintf("license.meta.force_online_after_sec (%d) exceeds max horizon (%d)", sec, forceOnlineMaxSec), nil)
	}
	abs := iat + sec
	return &abs, nil
}

func validateForceOnlineDeadline(candidate, iat int64, context string) error {
	if candidate < iat {
		return newError(CodeFingerprintRejected,
			fmt.Sprintf("%s (%d) is in the past (iat=%d)", context, candidate, iat), nil)
	}
	if candidate > iat+forceOnlineMaxSec {
		return newError(CodeFingerprintRejected,
			fmt.Sprintf("%s (%d) exceeds max horizon (%d)", context, candidate, iat+forceOnlineMaxSec), nil)
	}
	return nil
}

func resolveEntitlements(input IssueTokenInput) map[string]any {
	if input.Entitlements.Set {
		return input.Entitlements.Value
	}
	if input.License.Meta == nil {
		return nil
	}
	raw, ok := input.License.Meta["entitlements"]
	if !ok || raw == nil {
		return nil
	}
	m, ok := raw.(map[string]any)
	if !ok {
		return nil
	}
	return m
}

func unixSeconds(iso string) int64 {
	t, err := time.Parse(time.RFC3339Nano, iso)
	if err != nil {
		return 0
	}
	return t.Unix()
}

// toInt64 attempts to convert a JSON-decoded value to int64.
func toInt64(v any) (int64, bool) {
	switch x := v.(type) {
	case int:
		return int64(x), true
	case int64:
		return x, true
	case float64:
		if x == float64(int64(x)) {
			return int64(x), true
		}
		return 0, false
	default:
		return 0, false
	}
}

// singleKeyStore is a minimal KeyStore adapter that serves exactly one key.
// Used by IssueToken to satisfy KeyHierarchy.ImportSigningPrivate without a
// full storage tx.
type singleKeyStore struct {
	key *LicenseKey
}

// Put is unreachable — the single-key store is read-only for token issuance.
func (s *singleKeyStore) Put(_ LicenseKey) error {
	return fmt.Errorf("singleKeyStore.Put: unreachable")
}

// Get fetches the record.
func (s *singleKeyStore) Get(_ string) (*LicenseKey, error) {
	return nil, nil
}

// FindByKid finds by kid.
func (s *singleKeyStore) FindByKid(kid string) (*LicenseKey, error) {
	if kid == s.key.Kid {
		return s.key, nil
	}
	return nil, nil
}

// List lists the record.
func (s *singleKeyStore) List(_ KeyStoreFilter) ([]LicenseKey, error) {
	return []LicenseKey{*s.key}, nil
}

// Update updates the record.
func (s *singleKeyStore) Update(_ string, _ LicenseKey) error {
	return fmt.Errorf("singleKeyStore.Update: unreachable")
}
